src/pybind/mgr/dashboard/frontend/src/app/ceph/rgw/rgw-bucket-form/rgw-bucket-form.component.spec.ts

```diff
@@ -315,4 +317,80 @@ describe('RgwBucketFormComponent', () => {
       );
     });
   });
+
+  describe('object locking', () => {
+    it('should check lock enabled checkbox [mode=create]', () => {
+      component.createForm();
+      const control = component.bucketForm.get('lock_enabled');
+      expect(control.disabled).toBeFalsy();
+    });
+
+    it('should check lock enabled checkbox [mode=edit]', () => {
+      component.editing = true;
+      component.createForm();
+      const control = component.bucketForm.get('lock_enabled');
+      expect(control.disabled).toBeTruthy();
+    });
+
+    it('should have the "eitherDaysOrYears" error', () => {
+      formHelper.setValue('lock_enabled', true);
+      ['lock_retention_period_days', 'lock_retention_period_years'].forEach((name: string) => {
+        const control = component.bucketForm.get(name);
+        control.updateValueAndValidity();
+        expect(control.value).toBe(0);
+        expect(control.invalid).toBeTruthy();
+        formHelper.expectError(control, 'eitherDaysOrYears');
+      });
+    });
+
+    it('should have the "pattern" error [1]', () => {
+      formHelper.setValue('lock_enabled', true, true);
```

Comment:
As this and other parts of your test code are used multiple times it makes sense to use a function call to run these tests in order to easier spot the difference of the tests.

During my testing I just refactored it a bit to get a better overview:
```diff
diff --git a/src/pybind/mgr/dashboard/frontend/src/app/ceph/rgw/rgw-bucket-form/rgw-bucket-form.component.spec.ts b/src/pybind/mgr/dashboard/frontend/src/app/ceph/rgw/rgw-bucket-form/rgw-bucket-form.component.spec.ts
index 445a837ab6..0cb297c3f1 100644
--- a/src/pybind/mgr/dashboard/frontend/src/app/ceph/rgw/rgw-bucket-form/rgw-bucket-form.component.spec.ts
+++ b/src/pybind/mgr/dashboard/frontend/src/app/ceph/rgw/rgw-bucket-form/rgw-bucket-form.component.spec.ts
@@ -319,6 +319,35 @@ describe('RgwBucketFormComponent', () => {
   });
 
   describe('object locking', () => {
+    const setDaysAndYears = (fn: (name: string) => void) => {
+      ['lock_retention_period_days', 'lock_retention_period_years'].forEach(fn);
+    };
+
+    const expectPatternLockError = (value: string) => {
+      formHelper.setValue('lock_enabled', true, true);
+      setDaysAndYears((name: string) => {
+        formHelper.setValue(name, value);
+        formHelper.expectError(name, 'pattern');
+      });
+    };
+
+    const expectValidLockInputs = (enabled: boolean, mode: string, days: string, years: string) => {
+      formHelper.setValue('lock_enabled', enabled);
+      formHelper.setValue('lock_mode', mode);
+      formHelper.setValue('lock_retention_period_days', days);
+      formHelper.setValue('lock_retention_period_years', years);
+      [
+        'lock_enabled',
+        'lock_mode',
+        'lock_retention_period_days',
+        'lock_retention_period_years'
+      ].forEach((name) => {
+        const control = component.bucketForm.get(name);
+        expect(control.valid).toBeTruthy();
+        expect(control.errors).toBeNull();
+      });
+    };
+
     it('should check lock enabled checkbox [mode=create]', () => {
       component.createForm();
       const control = component.bucketForm.get('lock_enabled');
@@ -334,7 +363,7 @@ describe('RgwBucketFormComponent', () => {
 
     it('should have the "eitherDaysOrYears" error', () => {
       formHelper.setValue('lock_enabled', true);
-      ['lock_retention_period_days', 'lock_retention_period_years'].forEach((name: string) => {
+      setDaysAndYears((name: string) => {
         const control = component.bucketForm.get(name);
         control.updateValueAndValidity();
         expect(control.value).toBe(0);
@@ -344,53 +373,19 @@ describe('RgwBucketFormComponent', () => {
     });
 
     it('should have the "pattern" error [1]', () => {
-      formHelper.setValue('lock_enabled', true, true);
-      ['lock_retention_period_days', 'lock_retention_period_years'].forEach((name: string) => {
-        formHelper.setValue(name, '-1');
-        formHelper.expectError(name, 'pattern');
-      });
+      expectPatternLockError('-1');
     });
 
     it('should have the "pattern" error [2]', () => {
-      formHelper.setValue('lock_enabled', true, true);
-      ['lock_retention_period_days', 'lock_retention_period_years'].forEach((name: string) => {
-        formHelper.setValue(name, '1.2');
-        formHelper.expectError(name, 'pattern');
-      });
+      expectPatternLockError('1.2');
     });
 
     it('should have valid values [1]', () => {
-      formHelper.setValue('lock_enabled', true);
-      formHelper.setValue('lock_mode', 'Governance');
-      formHelper.setValue('lock_retention_period_days', '0');
-      formHelper.setValue('lock_retention_period_years', '1');
-      [
-        'lock_enabled',
-        'lock_mode',
-        'lock_retention_period_days',
-        'lock_retention_period_years'
-      ].forEach((name) => {
-        const control = component.bucketForm.get(name);
-        expect(control.valid).toBeTruthy();
-        expect(control.errors).toBeNull();
-      });
+      expectValidLockInputs(true, 'Governance', '0', '1');
     });
 
     it('should have valid values [2]', () => {
-      formHelper.setValue('lock_enabled', false);
-      formHelper.setValue('lock_mode', 'Compliance');
-      formHelper.setValue('lock_retention_period_days', '100');
-      formHelper.setValue('lock_retention_period_years', '0');
-      [
-        'lock_enabled',
-        'lock_mode',
-        'lock_retention_period_days',
-        'lock_retention_period_years'
-      ].forEach((name) => {
-        const control = component.bucketForm.get(name);
-        expect(control.valid).toBeTruthy();
-        expect(control.errors).toBeNull();
-      });
+      expectValidLockInputs(false, 'Compliance', '100', '0');
     });
   });
 });

```

Reply:
Done